apps/examples/src/examples/speech-bubble/SpeechBubble/helpers.tsx

```diff
@@ -1,211 +1,158 @@
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				br,
+				bl,
+			]
 			break
 		case 2:
-			modifiedSegments.splice(3, 0, ...createTailSegments('horizontal'))
+			// bottom
+			vertices = [
+				tl,
+				tr,
+				br,
+				new Vec2d(adjustedIntersection.x + offsetH, adjustedIntersection.y),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x - offsetH, adjustedIntersection.y),
+				bl,
+			]
 			break
 		case 3:
-			modifiedSegments = [...modifiedSegments, ...createTailSegments('vertical')]
+			// left
+			vertices = [
+				tl,
+				tr,
+				br,
+				bl,
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y + offsetV),
+				new Vec2d(handleInShapeSpace.x, handleInShapeSpace.y),
+				new Vec2d(adjustedIntersection.x, adjustedIntersection.y - offsetV),
+			]
 			break
 		default:
 			// eslint-disable-next-line no-console
-			console.log('default')
+			throw Error("no intersection found, this shouldn't happen")
 	}
 
-	return modifiedSegments
+	return vertices
 }
 
-export function getHandleIntersectionPoint({
-	w,
-	h,
-	handle,
-}: {
-	w: number
-	h: number
-	handle: TLHandle
-}) {
-	const offset = { horizontal: w / 10, vertical: h / 10 }
-	const handleVec = new Vec2d(handle.x, handle.y)
+export function getHandleIntersectionPoint(shape: SpeechBubbleShape) {
+	const { w, h, handles } = shape.props
+	const handleInShapeSpace = new Vec2d(handles.handle.x * w, handles.handle.y * h)
+
 	const center = new Vec2d(w / 2, h / 2)
-	const box = [new Vec2d(0, 0), new Vec2d(w, 0), new Vec2d(w, h), new Vec2d(0, h)]
+	const corners = [new Vec2d(0, 0), new Vec2d(w, 0), new Vec2d(w, h), new Vec2d(0, h)]
+	const segments = [
+		[corners[0], corners[1]],
+		[corners[1], corners[2]],
+		[corners[2], corners[3]],
+		[corners[3], corners[0]],
+	]
 
-	const result = checkIntersection(handleVec, center, box)
-	if (!result) return { intersection: null, offset: null, line: null }
-	const { result: intersection, line } = result
+	let segmentsIntersection: Vec2d | null = null
+	let intersectionSegment: Vec2d[] | null = null
 
-	// lines
-	///      0
-	//  _____________
-	//  |           |
-	// 3|           | 1
-	//  |           |
-	//  -------------
-	//        2
+	// If the point inside of the box's corners?
+	const insideShape = pointInPolygon(handleInShapeSpace, corners)
 
-	const intersectionVec = new Vec2d(intersection[0].x, intersection[0].y)
-	const lineCoordinates = {
-		0: { start: new Vec2d(0, 0), end: new Vec2d(w, 0) },
-		1: { start: new Vec2d(w, 0), end: new Vec2d(w, h) },
-		2: { start: new Vec2d(0, h), end: new Vec2d(w, h) },
-		3: { start: new Vec2d(0, 0), end: new Vec2d(0, h) },
-	}
+	// We want to be sure we get an intersection, so if the point is
+	// inside the shape, push it away from the center by a big distance
+	const pointToCheck = insideShape
+		? Vec2d.Add(handleInShapeSpace, Vec2d.Sub(handleInShapeSpace, center).uni().mul(1000000))
```

Comment:
You'll see this pattern a lot: `Vec2d.Add(start, Vec2d.Sub(end, start).uni().mul(distance)`.

It's saying, "find the normalized vector from start to end, then move in that direction by some distance". The resulting point will be `distance` away from `start` in the direction of `end`. Kind of like what you had with the angle, but IMO easier to read and recognize.

In this case, we want to handle that inside case by shoving that other point really far away.

Reply:
Before, the distance to move the point was calculated by using the slanted length of the shape, from top left to bottom right. Is using a really big number preferable?